### PR TITLE
Solve the random state serialization

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

New features
### PR changes

Others
### Describe

**Problem**:
Paddle's model training, the random state cannot be saved, resulting in hot start training, and the randomness cannot be completely restored.
**Solution**:
Support the random state serialization and using save the conditions.

### CODE REVIEW

**Review by zhiqiu — DISMISSED** [review]
LGTM

File: paddle/fluid/pybind/generator_py.cc [hunk 1]
```diff
@@ -39,7 +39,36 @@ void BindGenerator(py::module* m_ptr) {
...
+#if defined(PADDLE_WITH_CUDA) || defined(PADDLE_WITH_HIP)
+      // NOTE(shenliang03): Due to the inability to serialize mt19937_64
+      // type, resulting in a problem with precision under the cpu.
+      .def(py::pickle(
+          [](const phi::Generator::GeneratorState& s) {  // __getstate__
+            return py::make_tuple(s.device, s.current_seed, s.thread_offset);
+          },
+          [](py::tuple s) {  // __setstate__
+            if (s.size() != 3)
+              throw std::runtime_error("Invalid Random state!");
```
Comment (gongweibao) [inline paddle/fluid/pybind/generator_py.cc]: `Invalid Random state. Please check the format(device, current_seed, thread_offset)`.

File: python/paddle/fluid/tests/unittests/test_cuda_random_seed.py [hunk 2]
```diff
@@ -169,6 +171,42 @@ def test_gen_TruncatedNormal_initializer(self):
...
+        lt_dict = paddle.load(random_file)
+        st = lt_dict["random_state"]
+        paddle.set_cuda_rng_state(st)
+        x3 = paddle.randn([120], dtype="float32")
+
+        x1_np = x1.numpy()
+        x2_np = x2.numpy()
+        x3_np = x3.numpy()
+
+        if core.is_compiled_with_cuda():
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_cuda_random_seed.py]: Can we use `paddle.set_cuda_rng_state`  while not  `is_compiled_with_cuda`?

File: python/paddle/fluid/tests/unittests/test_cuda_random_seed.py [hunk 3]
```diff
@@ -169,6 +171,42 @@ def test_gen_TruncatedNormal_initializer(self):
...
+            print(">>>>>>> gaussian random dygraph state load/save >>>>>>>")
+            np.testing.assert_allclose(x1_np, x2_np, rtol=1e-05)
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_cuda_random_seed.py]: 1e-06 or higher? or compare with bit?

**Review by gongweibao — CHANGES_REQUESTED** [review]
Explain why we do this? Like what issue or problem to solve under what condition?

File: paddle/fluid/pybind/generator_py.cc (hunk 1 above)
Comment (ForFishes) [reply paddle/fluid/pybind/generator_py.cc]: done

File: python/paddle/fluid/tests/unittests/test_cuda_random_seed.py (hunk 2 above)
Comment (ForFishes) [reply python/paddle/fluid/tests/unittests/test_cuda_random_seed.py]: no, when not ``is_compiled_with_cuda ``,  ``paddle.set_cuda_rng_state`` is invalid.

File: python/paddle/fluid/tests/unittests/test_cuda_random_seed.py (hunk 3 above)
Comment (ForFishes) [reply python/paddle/fluid/tests/unittests/test_cuda_random_seed.py]: done

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**